File: apps/desktop/sidebar/gxserver-runtime/notification-feed.ts
/**
 * CDXC:Notifications 2026-09-11 WHY:
 * The native titlebar paints the bell and the notification panel from the
 * cached feed state this module posts over the bridge; it never issues an RPC
 * of its own. Row clicks and the jump keys come back here as one command
 * event, so marking a row read and focusing its session run on the same
 * `focusSession` path the sidebar uses, and the daemon stays the only owner of
 * read state.
 * SEE-ALSO: packages/shared/notification-feed/notification-feed-contract.ts,
 * server/src/notification_feed, apps/desktop/src/notification_feed.
 */
import {
  GPUI_SIDEBAR_NOTIFICATION_FEED_COMMAND_EVENT_NAME,
  GPUI_SIDEBAR_NOTIFICATION_FEED_STATE_MESSAGE_TYPE,
} from './constants';
import type { GpuiSidebarRuntime } from './core';
import { localGxserverSessionIdForSidebarSession } from './helpers/app-shot';
import { createGxserverPresentationProjectSessionId } from '@/packages/shared/gxserver-presentation-sidebar-projection';
import type { SidebarSessionItem } from '@/packages/shared/session-grid-contract';
import {
  NOTIFICATION_FEED_READ_ENDPOINT,
  NOTIFICATION_FEED_UPDATE_ENDPOINT,
  normalizeNotificationFeedState,
} from '@/packages/shared/notification-feed/notification-feed-contract';
import type {
  NotificationFeedItem,
  NotificationFeedState,
  NotificationFeedUpdateParams,
} from '@/packages/shared/notification-feed/notification-feed-contract';

let notificationRevealRequestId = Date.now();

const NOTIFICATION_FEED_COMMAND_ACTIONS = [
  'open',
  'dismiss',
  'markRead',
  'markUnread',
  'markAllRead',
  'clearAll',
  'jumpToLatestUnread',
  'deferAndJumpNext',
] as const;

type NotificationFeedCommandAction = (typeof NOTIFICATION_FEED_COMMAND_ACTIONS)[number];

function isNotificationFeedCommandAction(value: unknown): value is NotificationFeedCommandAction {
  return typeof value === 'string' && (NOTIFICATION_FEED_COMMAND_ACTIONS as readonly string[]).includes(value);
}

export interface GpuiSidebarRuntimeNotificationFeedMethods {
  refreshNotificationFeed(): Promise<void>;
  postNotificationFeedState(): void;
  updateNotificationFeed(params: NotificationFeedUpdateParams): Promise<NotificationFeedState | undefined>;
  openNotificationFeedItem(item: NotificationFeedItem): Promise<void>;
  jumpToLatestUnreadNotification(): Promise<void>;
  deferNotificationAndJumpNext(): Promise<void>;
  findSidebarSessionForNotificationFeedItem(sessionId: string): SidebarSessionItem | undefined;
  focusedLocalGxserverSessionId(): string | undefined;
  handleGpuiSidebarNotificationFeedCommand(event: Event): void;
}

export const gpuiSidebarRuntimeNotificationFeedMethods = {
  async refreshNotificationFeed(this: GpuiSidebarRuntime): Promise<void> {
    const client = this.client;
    if (!client) {
      return;
    }
    let state: NotificationFeedState;
    try {
      state = normalizeNotificationFeedState(await client.rpc<unknown>(NOTIFICATION_FEED_READ_ENDPOINT, {}));
    } catch {
      return;
    }
    this.notificationFeedState = state;
    this.postNotificationFeedState();
  },

  /** Deduplicated so a refetch that changed nothing does not wake the bridge. */
  postNotificationFeedState(this: GpuiSidebarRuntime): void {
    const state = this.notificationFeedState;
    const message = {
      items: state.items,
      ...(state.nextUnreadId ? { nextUnreadId: state.nextUnreadId } : {}),
      type: GPUI_SIDEBAR_NOTIFICATION_FEED_STATE_MESSAGE_TYPE,
      unreadCount: state.unreadCount,
    };
    const payload = JSON.stringify(message);
    if (payload === this.lastNotificationFeedStatePayload) {
      return;
    }
    this.lastNotificationFeedStatePayload = payload;
    window.webkit?.messageHandlers?.ghostexNativeHost?.postMessage(message);
  },

  async updateNotificationFeed(
    this: GpuiSidebarRuntime,
    params: NotificationFeedUpdateParams
  ): Promise<NotificationFeedState | undefined> {
    const client = this.client;
    if (!client) {
      return undefined;
    }
    let state: NotificationFeedState;
    try {
      state = normalizeNotificationFeedState(await client.rpc<unknown>(NOTIFICATION_FEED_UPDATE_ENDPOINT, params));
    } catch {
      return undefined;
    }
    this.notificationFeedState = state;
    this.postNotificationFeedState();
    return state;
  },

  /**
   * The feed stores raw daemon session ids while the sidebar addresses local
   * sessions by their presentation id (sometimes project-scoped), so the match
   * goes through the same id projection every other local lookup uses.
   */
  findSidebarSessionForNotificationFeedItem(
    this: GpuiSidebarRuntime,
    sessionId: string
  ): SidebarSessionItem | undefined {
    for (const group of this.latestGroups) {
      for (const session of group.sessions) {
        if (localGxserverSessionIdForSidebarSession(session) === sessionId) {
          return session;
        }
      }
    }
    return undefined;
  },

  focusedLocalGxserverSessionId(this: GpuiSidebarRuntime): string | undefined {
    const focusedSessionId = this.focusedSessionId;
    if (!focusedSessionId) {
      return undefined;
    }
    for (const group of this.latestGroups) {
      for (const session of group.sessions) {
        if (session.sessionId === focusedSessionId) {
          return localGxserverSessionIdForSidebarSession(session);
        }
      }
    }
    return undefined;
  },

  /**
   * CDXC:Notifications 2026-09-11 DECISION:
   * User: clicking a notification always takes you to that session and reveals it in the sidebar.
   * The row's raw ids are turned into the sidebar's project-scoped session id without requiring the row to be in the current snapshot, so a session hidden by a Space, a filter, or a stale snapshot still gets focused; the reveal request then selects its Space, clears filters, and scrolls to it once the row is published. The daemon drops rows for deleted sessions on read, so a missing session never strands a row.
   */
  async openNotificationFeedItem(this: GpuiSidebarRuntime, item: NotificationFeedItem): Promise<void> {
    const sidebarSessionId =
      this.findSidebarSessionForNotificationFeedItem(item.sessionId)?.sessionId ??
      createGxserverPresentationProjectSessionId(item.projectId, item.sessionId);
    await this.updateNotificationFeed({ action: 'markRead', notificationId: item.id });
    void this.focusSession(sidebarSessionId, { sessionId: sidebarSessionId, type: 'focusSession' });
    this.messageSource.postMessage({
      requestId: ++notificationRevealRequestId,
      sessionId: sidebarSessionId,
      type: 'revealSidebarSession',
    });
  },

  async jumpToLatestUnreadNotification(this: GpuiSidebarRuntime): Promise<void> {
    await this.refreshNotificationFeed();
    const nextUnreadId = this.notificationFeedState.nextUnreadId;
    const item = nextUnreadId ? this.notificationFeedState.items.find((entry) => entry.id === nextUnreadId) : undefined;
    if (item) {
      await this.openNotificationFeedItem(item);
    }
  },

  /**
   * "I am not done with this one; come back to it last." The focused session's
   * newest row goes to the back of the unread queue, then the jump opens
   * whatever the daemon now says is next, unless that is the same session.
   */
  async deferNotificationAndJumpNext(this: GpuiSidebarRuntime): Promise<void> {
    const sessionId = this.focusedLocalGxserverSessionId();
    const state = sessionId
      ? await this.updateNotificationFeed({ action: 'deferUnread', sessionId })
      : (await this.refreshNotificationFeed(), this.notificationFeedState);
    if (!state?.nextUnreadId) {
      return;
    }
    const item = state.items.find((entry) => entry.id === state.nextUnreadId);
    if (!item || item.sessionId === sessionId) {
      return;
    }
    await this.openNotificationFeedItem(item);
  },

  handleGpuiSidebarNotificationFeedCommand(this: GpuiSidebarRuntime, event: Event): void {
    const detail = (event as CustomEvent<unknown>).detail;
    if (!detail || typeof detail !== 'object') {
      return;
    }
    const { action, notificationId } = detail as { action?: unknown; notificationId?: unknown };
    if (!isNotificationFeedCommandAction(action)) {
      return;
    }
    const id = typeof notificationId === 'string' && notificationId.length > 0 ? notificationId : undefined;
    switch (action) {
      case 'open': {
        const item = id ? this.notificationFeedState.items.find((entry) => entry.id === id) : undefined;
        if (item) {
          void this.openNotificationFeedItem(item);
        }
        return;
      }
      case 'dismiss':
      case 'markRead':
      case 'markUnread':
        if (id) {
          void this.updateNotificationFeed({ action, notificationId: id });
        }
        return;
      case 'markAllRead':
      case 'clearAll':
        void this.updateNotificationFeed({ action });
        return;
      case 'jumpToLatestUnread':
        void this.jumpToLatestUnreadNotification();
        return;
      case 'deferAndJumpNext':
        void this.deferNotificationAndJumpNext();
        return;
    }
  },
};

export { GPUI_SIDEBAR_NOTIFICATION_FEED_COMMAND_EVENT_NAME };

File: packages/gx-core/src/notification_feed.rs
//! The notification feed as the host keeps it: the daemon's answer normalized into the one message
//! the bell and its panel read, and the command vocabulary the panel and the jump keys send.
//!
//! CDXC:Notifications 2026-09-25 WHY:
//! gxserver owns the rows, their read state and the jump order; a client only reads, normalizes and
//! asks for updates. This is the normalizer the old QuickJS runtime ran on every read
//! (`normalizeNotificationFeedState` in packages/shared/notification-feed/notification-feed-contract.ts),
//! moved here so the desktop, the web build and mobile keep the same lenient reading: a row with a
//! missing id, project, session, time or an unknown kind is dropped rather than failing the feed.
//!
//! SEE-ALSO: apps/desktop/src/app/gx_store/notifications/ (the desktop performs the reads),
//! apps/desktop/src/notification_feed/mod.rs (the bell, which parses this message strictly),
//! server/src/notification_feed/.

use serde_json::{Map, Value};

pub const NOTIFICATION_FEED_READ_ENDPOINT: &str = "/api/readNotificationFeed";
pub const NOTIFICATION_FEED_UPDATE_ENDPOINT: &str = "/api/updateNotificationFeed";

/// `type` of the state message the bell reads.
pub const NOTIFICATION_FEED_STATE_MESSAGE_TYPE: &str = "notificationFeedState";

const KINDS: [&str; 4] = ["finished", "needsInput", "bell", "custom"];

/// What the panel and the jump keys ask for.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum NotificationFeedCommand {
    Open,
    Dismiss,
    MarkRead,
    MarkUnread,
    MarkAllRead,
    ClearAll,
    JumpToLatestUnread,
    DeferAndJumpNext,
}

impl NotificationFeedCommand {
    pub fn from_action(action: &str) -> Option<Self> {
        Some(match action {
            "open" => Self::Open,
            "dismiss" => Self::Dismiss,
            "markRead" => Self::MarkRead,
            "markUnread" => Self::MarkUnread,
            "markAllRead" => Self::MarkAllRead,
            "clearAll" => Self::ClearAll,
            "jumpToLatestUnread" => Self::JumpToLatestUnread,
            "deferAndJumpNext" => Self::DeferAndJumpNext,
            _ => return None,
        })
    }

    /// The update endpoint's action for the commands that are one update and nothing else.
    pub fn update_action(self) -> Option<&'static str> {
        Some(match self {
            Self::Dismiss => "dismiss",
            Self::MarkRead => "markRead",
            Self::MarkUnread => "markUnread",
            Self::MarkAllRead => "markAllRead",
            Self::ClearAll => "clearAll",
            _ => return None,
        })
    }

    /// Whether the update names one row. `markAllRead` and `clearAll` take nothing, and a
    /// row command without a row id does nothing at all.
    pub fn takes_notification_id(self) -> bool {
        matches!(
            self,
            Self::Open | Self::Dismiss | Self::MarkRead | Self::MarkUnread
        )
    }
}

fn text<'a>(raw: &'a Map<String, Value>, key: &str) -> Option<&'a str> {
    raw.get(key).and_then(Value::as_str)
}

/// One row, or `None` when it lacks what a row needs. Key order is the TypeScript's.
fn normalize_item(value: &Value) -> Option<Value> {
    let raw = value.as_object()?;
    let non_empty = |key: &str| text(raw, key).filter(|value| !value.is_empty());
    let id = non_empty("id")?;
    let project_id = non_empty("projectId")?;
    let session_id = non_empty("sessionId")?;
    let created_at = non_empty("createdAt")?;
    let kind = non_empty("kind").filter(|kind| KINDS.contains(kind))?;
    let mut item = Map::new();
    if let Some(agent_name) = non_empty("agentName") {
        item.insert("agentName".into(), Value::from(agent_name));
    }
    item.insert("body".into(), Value::from(text(raw, "body").unwrap_or("")));
    item.insert("createdAt".into(), Value::from(created_at));
    item.insert("id".into(), Value::from(id));
    item.insert("kind".into(), Value::from(kind));
    item.insert("projectId".into(), Value::from(project_id));
    item.insert(
        "read".into(),
        Value::Bool(raw.get("read") == Some(&Value::Bool(true))),
    );
    item.insert("sessionId".into(), Value::from(session_id));
    item.insert(
        "subtitle".into(),
        Value::from(text(raw, "subtitle").unwrap_or("")),
    );
    item.insert(
        "title".into(),
        Value::from(text(raw, "title").unwrap_or("")),
    );
    Some(Value::Object(item))
}

/// The daemon's feed answer as the state message the bell reads:
/// `{ items, nextUnreadId?, type, unreadCount }`. Anything that is not an object is the empty
/// feed, as in the TypeScript, which posted it.
pub fn notification_feed_state_message(result: &Value) -> Value {
    let raw = result.as_object();
    let items: Vec<Value> = raw
        .and_then(|raw| raw.get("items"))
        .and_then(Value::as_array)
        .map(|items| items.iter().filter_map(normalize_item).collect())
        .unwrap_or_default();
    let unread_count = raw
        .and_then(|raw| raw.get("unreadCount"))
        .and_then(Value::as_f64)
        .filter(|count| count.is_finite() && *count >= 0.0)
        .map(|count| count.floor() as u64)
        .unwrap_or_else(|| {
            items
                .iter()
                .filter(|item| item["read"] != Value::Bool(true))
                .count() as u64
        });
    let next_unread_id = raw
        .and_then(|raw| text(raw, "nextUnreadId"))
        .filter(|id| !id.is_empty());
    let mut message = Map::new();
    message.insert("items".into(), Value::Array(items));
    if let Some(next_unread_id) = next_unread_id {
        message.insert("nextUnreadId".into(), Value::from(next_unread_id));
    }
    message.insert(
        "type".into(),
        Value::from(NOTIFICATION_FEED_STATE_MESSAGE_TYPE),
    );
    message.insert("unreadCount".into(), Value::from(unread_count));
    Value::Object(message)
}

/// The row a jump opens, from a state message: the daemon's `nextUnreadId`, unless that row is
/// the session the user is deferring away from.
pub fn notification_feed_jump_target<'a>(
    message: &'a Value,
    deferred_session_id: Option<&str>,
) -> Option<&'a Value> {
    let next = message.get("nextUnreadId")?.as_str()?;
    let item = message
        .get("items")?
        .as_array()?
        .iter()
        .find(|item| item["id"].as_str() == Some(next))?;
    if deferred_session_id.is_some() && item["sessionId"].as_str() == deferred_session_id {
        return None;
    }
    Some(item)
}
